src/Grpc.AspNetCore.Server.Reflection/Internal/ReflectionGrpcServiceActivator.cs

```diff
@@ -0,0 +1,113 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Grpc.Core;
+using Grpc.Reflection;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Grpc.AspNetCore.Server.Reflection.Internal
+{
+    internal class ReflectionGrpcServiceActivator : IGrpcServiceActivator<ReflectionServiceImpl>
+    {
+        private readonly RouteOptions _routeOptions;
+        private readonly ILogger<ReflectionGrpcServiceActivator> _logger;
+
+        private ReflectionServiceImpl _instance;
+
+        public ReflectionGrpcServiceActivator(IOptions<RouteOptions> routeOptions, ILoggerFactory loggerFactory)
+        {
+            _routeOptions = routeOptions.Value;
+            _logger = loggerFactory.CreateLogger<ReflectionGrpcServiceActivator>();
+        }
+
+        public ReflectionServiceImpl Create()
+        {
+            if (_instance == null)
+            {
+                var grpcEndpointMetadata = _routeOptions.EndpointDataSources
+                    .SelectMany(ds => ds.Endpoints)
+                    .Select(ep => ep.Metadata.GetMetadata<GrpcMethodMetadata>())
+                    .Where(m => m != null)
+                    .ToList();
+
+                var serviceTypes = grpcEndpointMetadata.Select(m => m.ServiceType).Distinct().ToList();
+
+                var serviceDescriptors = new List<Google.Protobuf.Reflection.ServiceDescriptor>();
+
+                foreach (var serviceType in serviceTypes)
+                {
+                    var baseType = GetServiceBaseType(serviceType);
+                    var definitionType = baseType?.DeclaringType;
+
+                    var descriptorPropertyInfo = definitionType?.GetProperty("Descriptor", BindingFlags.Public | BindingFlags.Static);
+                    if (descriptorPropertyInfo != null)
+                    {
+                        var serviceDescriptor = descriptorPropertyInfo.GetValue(null) as Google.Protobuf.Reflection.ServiceDescriptor;
+                        if (serviceDescriptor != null)
+                        {
+                            serviceDescriptors.Add(serviceDescriptor);
+                            continue;
+                        }
```

Comment:
Shouldn't it be an error if we can't get the descriptor value?

Reply:
Sorry, not an error, but at least we should log it.

Reply:
You want it logged like it is already logged on line 74? 😋